go-pkg/ds/permission.go

```diff
@@ -1,27 +1,94 @@
 package ds
 
 import (
+	"crypto/ecdsa"
+	"errors"
+
+	ethereum "github.com/ethereum/go-ethereum/crypto"
+	"github.com/jinfwhuang/ds-toolkit/go-pkg/encrypt"
 	protods "github.com/jinfwhuang/ds-toolkit/proto/ds"
 )
 
 // Do I have permission to the DataBlob
-func (u *User) checkPerm(data *protods.DataBlob) bool {
-	panic("not implemented")
+func checkPerm(data *protods.DataBlob, pubKey *ecdsa.PublicKey) bool {
+	pubKeyBytes := ethereum.CompressPubkey(pubKey)
+	_, err := findUserKey(data.Keys, pubKeyBytes)
+	return err == nil
 }
 
 // Get the decrypted data in EncryptedData
-func (u *User) extractData(data *protods.DataBlob) []byte {
-	panic("not implemented")
+func extractData(data *protods.DataBlob, privKey *ecdsa.PrivateKey) ([]byte, error) {
+	pubKeyBytes := ethereum.CompressPubkey(&privKey.PublicKey)
+	userKey, err := findUserKey(data.Keys, pubKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+	dataKey, err := recoverHiddenDataKey(userKey, privKey.D.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	decryptedData, err := encrypt.Decrypt(dataKey, data.Iv, data.EncryptedData)
+	if err != nil {
+		return nil, errors.New("failed to decrypt data")
+	}
+
+	// AES blocks are padded, we need to get rid of the padding
+	unpaddedDecryptedData := removeBlockCipherPadding(decryptedData)
+
+	return unpaddedDecryptedData, nil
 }
 
 // User add key to an existing Blob and creat a new DataBlob
-func (u *User) addKey(blob *protods.DataBlob, newDataOwner *User) *protods.DataBlob {
-	panic("not implemented")
+func addKey(blob *protods.DataBlob, newPubKey *ecdsa.PublicKey, privKey *ecdsa.PrivateKey) (*protods.DataBlob, error) {
+	pubKeyBytes := ethereum.CompressPubkey(&privKey.PublicKey)
+	userKey, err := findUserKey(blob.Keys, pubKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+	dataKey, err := recoverHiddenDataKey(userKey, privKey.D.Bytes())
+	if err != nil {
+		return nil, err
+	}
```

Comment:
This looks like repeated code.

Reply:
Indeed... Overall I don't like much the error handling in golang. I found myself
```if err != nil {
// handle error
}```
to be the best one as I make the function fail as early as possible and here both functions are internal one, so I already handled the error in some way in both `findUserKey` and `recoverHiddenDataKey`.